core/drivers/imx/dcp/dcp.c

```diff
@@ -0,0 +1,776 @@
+static struct dcp_align_buf hw_context_buffer;
+
+static struct mutex lock_channel[DCP_NB_CHANNELS] = {
+	[DCP_CHANN0] = MUTEX_INITIALIZER,
+	[DCP_CHANN1] = MUTEX_INITIALIZER,
+	[DCP_CHANN2] = MUTEX_INITIALIZER,
+	[DCP_CHANN3] = MUTEX_INITIALIZER,
+};
+
+static const struct dcp_hashalg hash_alg[2] = {
+	[DCP_SHA1] = {
+		.type = DCP_CONTROL1_HASH_SELECT_SHA1,
+		.size = TEE_SHA1_HASH_SIZE,
+	},
+	[DCP_SHA256] = {
+		.type = DCP_CONTROL1_HASH_SELECT_SHA256,
+		.size = TEE_SHA256_HASH_SIZE,
+	},
+};
+
+/*
+ * Enable/disable DCP clock.
+ *
+ * @enable   Enable the clock if true, disable if false.
+ */
+static void dcp_clk_enable(bool enable)
+{
+	vaddr_t ccm_base = core_mmu_get_va(CCM_BASE, MEM_AREA_IO_SEC);
+	uint32_t clock_except = cpu_spin_lock_xsave(&clock_spinlock);
+
+	if (enable) {
+		if (clk_refcount > 0) {
+			clk_refcount++;
+			goto out;
+		} else {
+			clk_refcount++;
+			io_setbits32(ccm_base + CCM_CCGR0, DCP_CLK_ENABLE_MASK);
+		}
+	} else {
+		assert(clk_refcount != 0);
+
+		clk_refcount--;
+		if (clk_refcount > 0)
+			goto out;
+		else
+			io_clrbits32(ccm_base + CCM_CCGR0, DCP_CLK_ENABLE_MASK);
+	}
+out:
+	cpu_spin_unlock_xrestore(&clock_spinlock, clock_except);
+}
+
+/*
+ * Lock the given channel with a mutex.
+ *
+ * @chan   DCP channel to lock
+ */
+static TEE_Result dcp_lock_known_channel(enum dcp_channel chan)
+{
+	if (mutex_trylock(&lock_channel[chan]))
+		return TEE_SUCCESS;
+	else
+		return TEE_ERROR_BUSY;
+}
+
+/*
+ * Lock a DCP channel
+ *
+ * @channel    Pointer on operation channel parameter
+ */
+static TEE_Result dcp_lock_channel(enum dcp_channel *channel)
+{
+	TEE_Result ret = TEE_ERROR_BUSY;
+	enum dcp_channel chan = DCP_CHANN0;
+
+	for (chan = DCP_CHANN0; chan < DCP_NB_CHANNELS; chan++) {
+		ret = dcp_lock_known_channel(chan);
+		if (ret == TEE_SUCCESS) {
+			*channel = chan;
+			return ret;
+		}
+	}
+
+	EMSG("All channels are busy");
+
+	return ret;
+}
+
+/*
+ * Unlock the given channel.
+ *
+ * @chan   DCP channel to unlock
+ */
+static void dcp_unlock_channel(enum dcp_channel chan)
+{
+	mutex_unlock(&lock_channel[chan]);
+}
+
+/*
+ * Start the DCP operation.
+ *
+ * @dcp_data   Structure containing dcp_descriptor configuration and channel to
+ *	       use.
+ */
+static TEE_Result dcp_run(struct dcp_data *dcp_data)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	unsigned int timeout = 0;
+	uint32_t val = 0;
+
+	dcp_data->desc.next = 0;
+	cache_operation(TEE_CACHEFLUSH, &dcp_data->desc,
+			sizeof(dcp_data->desc));
+
+	/* Enable clock if it's not done */
+	dcp_clk_enable(true);
+
+	/* Clear DCP_STAT IRQ field for the channel used by the operation */
+	io_clrbits32(dcp_base + DCP_STAT, BIT32(dcp_data->channel));
+
+	/* Clear CH_N_STAT to clear IRQ and error codes */
+	io_write32(dcp_base + DCP_CH_N_STAT(dcp_data->channel), 0x0);
+
+	/* Update descriptor structure to be processed for the channel */
+	io_write32(dcp_base + DCP_CH_N_CMDPTR(dcp_data->channel),
+		   virt_to_phys(&dcp_data->desc));
```

Comment:
I guess this means that hardware and software must have the same layout of `struct dcp_descriptor`.
How does that work since there's a few `void *` in that struct?

Reply:
```void *``` are 32 bits on platforms featuring DCP.
I can make it obvious and set it to ```uint32_t```

Reply:
Please do, it could be worth noting that those fields hold physical addresses too.

Reply:
Replaced with ```uint32_t``` and added a comment mentioning physical addresses.